server/src/main/java/org/opensearch/action/search/TransportGetAllPitsAction.java

```diff
@@ -8,79 +8,31 @@
 
-import java.io.IOException;
-import java.util.List;
-
 /**
- * Transport action to get all active PIT contexts across all nodes
+ * Transport action to get all active PIT contexts across the cluster
  */
-public class TransportGetAllPitsAction extends TransportNodesAction<
-    GetAllPitNodesRequest,
-    GetAllPitNodesResponse,
-    GetAllPitNodeRequest,
-    GetAllPitNodeResponse> {
-    private final SearchService searchService;
+public class TransportGetAllPitsAction extends HandledTransportAction<GetAllPitNodesRequest, GetAllPitNodesResponse> {
+    private final PitService pitService;
 
     @Inject
-    public TransportGetAllPitsAction(
-        ThreadPool threadPool,
-        ClusterService clusterService,
-        TransportService transportService,
-        ActionFilters actionFilters,
-        SearchService searchService
-    ) {
-        super(
-            GetAllPitsAction.NAME,
-            threadPool,
-            clusterService,
-            transportService,
-            actionFilters,
-            GetAllPitNodesRequest::new,
-            GetAllPitNodeRequest::new,
-            ThreadPool.Names.SAME,
-            GetAllPitNodeResponse.class
-        );
-        this.searchService = searchService;
-    }
-
-    @Override
-    protected GetAllPitNodesResponse newResponse(
-        GetAllPitNodesRequest request,
-        List<GetAllPitNodeResponse> getAllPitNodeRespons,
-        List<FailedNodeException> failures
-    ) {
-        return new GetAllPitNodesResponse(clusterService.getClusterName(), getAllPitNodeRespons, failures);
-    }
-
-    @Override
-    protected GetAllPitNodeRequest newNodeRequest(GetAllPitNodesRequest request) {
-        return new GetAllPitNodeRequest();
-    }
-
-    @Override
-    protected GetAllPitNodeResponse newNodeResponse(StreamInput in) throws IOException {
-        return new GetAllPitNodeResponse(in);
+    public TransportGetAllPitsAction(ActionFilters actionFilters, TransportService transportService, PitService pitService) {
+        super(GetAllPitsAction.NAME, transportService, actionFilters, in -> new GetAllPitNodesRequest(in));
+        this.pitService = pitService;
     }
 
-    /**
-     * This retrieves all active PITs in the node
-     */
-    @Override
-    protected GetAllPitNodeResponse nodeOperation(GetAllPitNodeRequest request) {
-        GetAllPitNodeResponse nodeResponse = new GetAllPitNodeResponse(
-            transportService.getLocalNode(),
-            searchService.getAllPITReaderContexts()
-        );
-        return nodeResponse;
+    protected void doExecute(Task task, GetAllPitNodesRequest request, ActionListener<GetAllPitNodesResponse> listener) {
+        // If security plugin intercepts the request, it'll replace all PIT IDs with permitted PIT IDs
+        if (request.getGetAllPitNodesResponse() != null) {
+            listener.onResponse(request.getGetAllPitNodesResponse());
+        } else {
+            pitService.getAllPits(listener);
+        }
```

Comment:
Ensure we handle cases for non-null but empty cases and handle them appropriately at security plugin

Reply:
Yes I have tested them locally, all are working as expected, will push the changes and share the final pr.